zcash_primitives/src/transaction/components/sapling/builder.rs

```diff
@@ -513,18 +520,100 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
             ));
         }
 
-        // Add a binding signature if needed
-        let binding_sig =
-            if tx_metadata.spend_indices.is_empty() && tx_metadata.output_indices.is_empty() {
+        if tx_metadata.spend_indices.is_empty() && tx_metadata.output_indices.is_empty() {
+            Ok(None)
+        } else {
+            let spend_sigs = spend_sigs
+                .into_iter()
+                .collect::<Option<Vec<Signature>>>()
+                .unwrap_or_default();
+
+            let binding_sig = prover
+                .binding_sig(ctx, self.value_balance, sighash_bytes)
+                .map_err(|_| Error::BindingSig)?;
+
+            Ok(Some((spend_sigs, binding_sig)))
+        }
+    }
+}
+
+#[cfg(any(test, feature = "test-dependencies"))]
+pub mod testing {
+    use proptest::collection::vec;
+    use proptest::prelude::*;
+    use rand::{rngs::StdRng, SeedableRng};
+
+    use crate::{
+        consensus::{
+            testing::{arb_branch_id, arb_height},
+            TEST_NETWORK,
+        },
+        merkle_tree::{testing::arb_commitment_tree, IncrementalWitness},
+        sapling::{
+            prover::{mock::MockTxProver, TxProver},
+            testing::{arb_node, arb_note, arb_positive_note_value},
+            Diversifier,
+        },
+        transaction::components::{
+            amount::MAX_MONEY,
+            sapling::{Authorized, Bundle},
+        },
+        zip32::testing::arb_extended_spending_key,
+    };
+
+    use super::SaplingBuilder;
+
+    prop_compose! {
+        fn arb_bundle()(n_notes in 1..30usize)(
+            extsk in arb_extended_spending_key(),
+            spendable_notes in vec(
+                arb_positive_note_value(MAX_MONEY as u64 / 10000).prop_flat_map(arb_note),
+                n_notes
+            ),
+            commitment_trees in vec(
+                arb_commitment_tree(arb_node()).prop_map(
```

Comment:
Technically, the tree needs to contain at least as many notes as are being spent, so this might be useful as an `arb_nonempty_commitment_tree(spendable_notes.len())`.